source/common/router/rds_impl.cc

```diff
@@ -306,11 +306,19 @@ void RdsRouteConfigProviderImpl::requestVirtualHostsUpdate(
     std::weak_ptr<Http::RouteConfigUpdatedCallback> route_config_updated_cb) {
   auto alias =
       VhdsSubscription::domainNameToAlias(config_update_info_->routeConfigName(), for_domain);
-  factory_context_.dispatcher().post([this, alias, &thread_local_dispatcher,
-                                      route_config_updated_cb]() -> void {
-    subscription_->updateOnDemand(alias);
-    config_update_callbacks_.push_back({alias, thread_local_dispatcher, route_config_updated_cb});
-  });
+  // We use weak pointers here as RdsRouteConfigProviderImpl instance could go away before the
+  // dispatcher had a chance to execute the callback.
+  factory_context_.dispatcher().post(
+      [sub = std::weak_ptr<RdsRouteConfigSubscription>(subscription_),
```

Comment:
I think you only need the config_cbs to help decide if the config provider is gone.

Reply:
Yes, but it's dependent on the position of the member var in the source file. I figured this is the safest?

Reply:
I dont get it: as you pointed out in the other comment, the RDS provider is guaranteed to be allocate/deallocated on main thread. so if we can lock config_cbs here, that means the subscription is definitely alive. this lambda is run in main thread, where deallocation of RDS provider happens in sequence with this lambda.

Reply:
You are right, fixed

Reply:
sorry, I think you want to check the config_cbs_ actually. 
Subscription is shared between routeconfigProviders, there could be a case where this route provider could be gone, but the subscription stays.